### DISCUSSION TITLE
RFC 23 - Consolidate Configurations

### DISCUSSION THREAD

**Discussion by hikchoi** [dendronhq/dendron#1260] [RFCs/Ideas]
This is a discussion thread for RFC 23 - Consolidate Configurations

A Pull request for this RFC can be found here: https://github.com/dendronhq/dendron-site/pull/175

**Comment by SeriousBug** [comment]
> If a setting affects multiple top level groups, they should be in each group

I like this, but I have one concern: following the same example, if a user just wants to enable `mermaid` everywhere what do they do? Do they have to configure it manually under each group it appears?

> prefer: `export enum LookupSelectionTypeEnum { "selectionExtract" = "selectionExtract" ...`

A small question/recommendation: should we use more natural-language identifiers here? `"selectionExtract" = "selection extract"` might allow us to present a slightly better UI.

**Reply by SeriousBug** [reply]
In case Github is butchering it, there's a space between "selection" and "extract" in my last sentence.

**Reply by hikchoi** [reply]
> I like this, but I have one concern: following the same example, if a user just wants to enable mermaid everywhere what do they do? Do they have to configure it manually under each group it appears?

Primarily, yes. If a user wants to do this by directly editing `dendron.yml`, they have to go under each group and set them to be true. But if the ability to enabling a feature in multiple places at once is deemed necessary, the `Dendron: Configure` web ui should provide a way to do this.

The immediate thing that comes to mind is aggregating all configurations that exist in multiple places in a (logically) separate group and presenting them in the configure UI.

Another possibility is to let users search and filter through settings in the configure UI so that if you type `mermaid`, you are presented with all configs that are `mermaid`, and only those.

The main point here is that presentation should be agnostic to the underlying structure of the configuration namespace. and vice versa. The need to present some set of configs in a particular way should not affect how the namespace is structured.

> A small question/recommendation: should we use more natural-language identifiers here?

This is a good point, and I should update so that the RFC goes into this more.

I think what you are suggesting could be done if we introduce labels and descriptions that are attached to each choices in the configuration. (also links back to my point of structure agnostic presentation). We can keep the existing string literals internally (and in `dendron.yml`), and when we present it in the configure UI we can use a natural language description of the identifier for a better user experience.

**Reply by kevinslin** [reply]
in terms of global config, one way to keep the convenience with the separation is by introducing a `global` block. this would set the default for all blocks mermaid was mentioned unless overridden by a specific section

```yml
global:
  mermaid: true
...
```

**Comment by hikchoi** [comment]
I've started working on this, and you can see a few namespaces I've added as per the RFC in the draft PR here: https://github.com/dendronhq/dendron/pull/1295

I've added a way to map the possible config choices to a more user friendly description here as well so let me know if this makes sense to you @SeriousBug
